instrumentation/messaging/src/main/java/brave/messaging/MessagingTracing.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2013-2019 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.messaging;
+
+import brave.Span;
+import brave.Tracing;
+import brave.propagation.TraceContext;
+import brave.propagation.TraceContextOrSamplingFlags;
+
+public class MessagingTracing {
+
+  public static MessagingTracing create(Tracing tracing) {
+    return newBuilder(tracing).build();
+  }
+
+  public static MessagingTracing.Builder newBuilder(Tracing tracing) {
+    return new MessagingTracing.Builder(tracing);
+  }
+
+  final Tracing tracing;
+  final MessagingConsumerParser consumerParser;
+  final MessagingProducerParser producerParser;
+
+  MessagingTracing(Builder builder) {
+    this.tracing = builder.tracing;
+    this.consumerParser = builder.consumerParser;
+    this.producerParser = builder.producerParser;
+  }
+
+  public Tracing tracing() {
+    return tracing;
+  }
+
+  public MessagingProducerParser producerParser() {
+    return producerParser;
+  }
+
+  public MessagingConsumerParser consumerParser() {
+    return consumerParser;
+  }
+
+  public <Chan, Msg> Span nextSpan(ChannelAdapter<Chan> channelAdapter,
```

Comment:
@jeqo I think we mentioned concerns like this in spring-cloud/spring-cloud-sleuth#1501 and openzipkin/brave#1047

probably we could resync with what is going on here as we spent quite a while on this topic (clearing headers) a few times.